### ISSUE TITLE
Angular 5 | Modal not opening

### ISSUE THREAD

**Issue by darthsitthiander** [sherweb/ng2-materialize#241]
There seems to be a bug when using a modal the programmatical way. When ng2-materialize is not installed as symbolic link but normally with npm, the button click event will generate the following error:

> ERROR TypeError: Cannot read property 'length' of undefined
    at MzInjectionService.getRootViewContainer (webpack-internal:///../../../../ng2-materialize/dist/shared/injection/injection.service.js:64)

I made a repo where I rewrote your demo-app to retrieve ng2-materialize normally and the error can be reproduced:

https://github.com/darthsitthiander/demo-app

**Comment by jfcere** [comment]
Hi @darthsitthiander, what a coincidence, I answered the same question this morning :)

There is an issue with Angular 5 with `MzInjectionService (see https://github.com/sherweb/ng2-materialize/issues/160#issuecomment-343997202).

There might be a workaround ... you would need to import `MzInjectionService` from `ng2-materialize/dist/shared` and use [setRootViewContainer](https://github.com/sherweb/ng2-materialize/blob/master/src/app/shared/injection/injection.service.ts#L67) method to set another container for the modal to be injected into so it does not try to fallback to `_rootComponents` private property that has been renamed in Angular 5.

**Comment by darthsitthiander** [comment]
Hey @jfcere! Wow, thanks for the quick reply. I've been struggling with this for a couple of hours now, so help is really appreciated. Just tried your approach and I'm trying to wrap my head around setRootViewContainer. The way I have my code now is
`import { Component, OnInit } from '@angular/core';
import { MzModalService } from 'ng2-materialize';
import { LoginComponent } from '../modals/login/login.component';
import { MzInjectionService } from 'ng2-materialize/dist/shared'

  constructor(
    private injectionService: MzInjectionService,
    private modalService: MzModalService) {}

  ngOnInit() {
    $(document).ready(function(){
      // the "href" attribute of the modal trigger must specify the modal ID that wants to be triggered
      $('.modal').modal();
    });
  }

public openServiceModal() {
    this.injectionService.setRootViewContainer(LoginComponent);
    this.modalService.open(LoginComponent);
  }`

I am getting the following error:

> GuesttoolbarComponent.html:13 ERROR TypeError: Cannot read property 'rootNodes' of undefined
    at MzInjectionService.getComponentRootNode

But I'm sure this has to do with my use of the above mentioned function.

**Comment by jfcere** [comment]
@darthsitthiander yeah I noticed I was a little "vague" on my example.

I noticed the changes I need to do to make it compatible with Angular 5 which I'm going to work in the next week.

But in the meenwhile the workaround is not pretty as you need to build an object what will "mimic" the `ComponentRef` that is supposed to be used...

```typescript
import { Component } from '@angular/core';
import { MzModalService } from 'ng2-materialize';
import { LoginComponent } from '../modals/login/login.component';
import { MzInjectionService } from 'ng2-materialize/dist/shared'

...

constructor(
  private injectionService: MzInjectionService,
  private modalService: MzModalService,
) {
  this.injectionService.setRootViewContainer({
    hostView: {
      rootNodes: [document.body],
    },
  });
}

openServiceModal() {
  this.modalService.open(LoginComponent);
}
```
Note that as the service is a singleton, you could assign the container through `setRootViewContainer` only once (lets say in your `AppComponent` constructor) to avoid to do this all over you application. This way when it is going be fixed for Angular 5 you will only have to remove the workaround at this unique place.

**Comment by darthsitthiander** [comment]
Thanks for the reply! I tried mentioned approach yet to no avail. First I had to add my modal component LoginComponent as entryComponent in the app.module.ts. Then I got error messages 

> GuesttoolbarComponent.html:13 ERROR TypeError: Cannot read property 'onClose' of undefined // as well as 'open'

Basically saying the componentRef couldn't be found.

I'll stick with the web version for now and implement this as soon as it's fixed. Thanks for the help and work on this. I'm looking forward to the upcoming release.

**Comment by jfcere** [comment]
I am not totaly sure to understand the issue with `onClose` and `open`, if you feel like it, you can provide me a link to your repository so I can take a look at it.

Another solution is to rollback to latest version of angular 4.

**Comment by darthsitthiander** [comment]
Hmm, I updated the above mentioned repo and it's working fine there:

https://github.com/darthsitthiander/demo-app

So the problem seems to be somewhere else in my application. I'll get back to you once I've found it.

**Comment by darthsitthiander** [comment]
Got it working now. I had to remove the ModalComponent declaration from the modal.module.ts and move it in the app.module.ts. I think the way I imported components and handled the injections had gone totally wrong somewhere, but I'm not investigating further, except if I can help somehow.

Thanks again!

**Comment by jfcere** [comment]
Anytime! I'll keep the issue open until I do the fix.

**Comment by jfcere** [comment]
@darthsitthiander I guess I never mentioned it but the fix has been released since [v1.7.0](https://github.com/sherweb/ng2-materialize/releases/tag/v1.7.0)

**Comment by darthsitthiander** [comment]
@jfcere Awesome, I'll check it out in the coming days. Thanks for the heads up!
